pkg/trait/pod_test.go

```diff
@@ -138,6 +150,9 @@ func createPodTest(podSpecTemplate string) (*podTrait, *Environment, *appsv1.Dep
 							},
 						},
 					},
+					SecurityContext: &corev1.PodSecurityContext{
```

Comment:
probably this part is not right. if you've implemented this feature right, PodTrait should update the deployment and you don't need to manually define them here. why do we need this part?

Reply:
If you see the previous lines on that file, it is composing a Deployment by hand, not using one coming from an integration.  All the tests on that file use a `corev1.PodTemplateSpec` spec, not the one `PodTemplateSpec` from `integration_types.go` (the one I changed).

I was debugging here on an Openshift instance, and as far as I can tell, on the POST client call where the value is sent to the Kubernetes API the value is present, but not on the return from the API.

Reply:
Figured out. The problem was with an outdated `resources.go` file.

Reply:
@ammachado But look at `(*podTrait).Appy()`:
https://github.com/apache/camel-k/blob/main/pkg/trait/pod.go#L65
It reads the `PodTemplateSpec` from `Integration` and does the job of applying the template changes to the `Deployment`. If it's hardcoded in the `Deployment` in the test settings, this step is not tested here.
Am I still missing something?

Reply:
It was missing my changes because I had an outdated `pkg/resources/resources.go`.  My tests are passing, but the build is still broken.

Reply:
The code you've mentioned in `pkg/trait/pod_test.go` is the "expected" deployment after applying the `Integration.Spec.PodTemplate.Spec`.

The code you've mentioned on `pkg/trait/pod.go#L65` uses the PodTemplate that is resolved on https://github.com/apache/camel-k/blob/main/pkg/cmd/run.go#L583. That code wasn't changed, it was not needed.

Just adding the `SecurityContext` field to the `PodSpec` template on `integration_types.go` made it work.  No other changes were needed.

Then a few test cases were added, and the CRDs and deepcopy code regenerated.

Reply:
A change I'd like to make is to make all fields of the `PodSpec` optional.  Today the field `Containers` is not optional (the only one).  I have to declare an empty `containers: []` on my pod template string to get accepted on `kamel run`.

Reply:
> The code you've mentioned in pkg/trait/pod_test.go is the "expected" deployment after applying the Integration.Spec.PodTemplate.Spec.

Hmm, I don't think so... Looking at the code `pod_test.go`, the returned template from `testPodTemplateSpec()` function appears in the "actual" side of assertions in all of the tests. Where is it that the deployment returned from `createPodTest()` function is used as the "expected" value in an assertion?

Reply:
Btw, you can ignore the test failure in the `openshift` workflow, which is just a known flaky test.
Now your code is ok, except only one thing that I still think the way `pod_test.go` is set up is not correct.

Reply:
> The code you've mentioned on pkg/trait/pod.go#L65 uses the PodTemplate that is resolved on https://github.com/apache/camel-k/blob/main/pkg/cmd/run.go#L583. That code wasn't changed, it was not needed.

Yeah, and the behaviour from `kamel run` is mimicked here
https://github.com/apache/camel-k/blob/main/pkg/trait/pod_test.go#L95-L98
so you can expect the same behaviour from the test code. That's why you wouldn't need an explicit addition of `SecurityContext` to deployment here.

Reply:
> > The code you've mentioned in pkg/trait/pod_test.go is the "expected" deployment after applying the Integration.Spec.PodTemplate.Spec.
> 
> Hmm, I don't think so... Looking at the code `pod_test.go`, the returned template from `testPodTemplateSpec()` function appears in the "actual" side of assertions in all of the tests. Where is it that the deployment returned from `createPodTest()` function is used as the "expected" value in an assertion?

Haven't noticed that before, but it is the expected case. The first time I've looked at that code I thought that the `Spec` on line 124 should come from the `podSpecTemplate` on line 24.  I changed it back then, but some other tests were failing, because they're expecting the container "integration" to be available (it is default container name for the Camel-K runtime), but it is created later during the `kamel run` process, where it applies the `container` trait.

This code is just testing the parsing of pod template strings.

Reply:
> > The code you've mentioned on pkg/trait/pod.go#L65 uses the PodTemplate that is resolved on https://github.com/apache/camel-k/blob/main/pkg/cmd/run.go#L583. That code wasn't changed, it was not needed.
> 
> Yeah, and the behaviour from `kamel run` is mimicked here https://github.com/apache/camel-k/blob/main/pkg/trait/pod_test.go#L95-L98 so you can expect the same behaviour from the test code. That's why you wouldn't need an explicit addition of `SecurityContext` to deployment here.

Figured out what you said.  I'm sorry for the misunderstanding.  What's happening is that the `yaml.Unmarshal` on line 109 is initializing the field `SecurityContext` on the `podSpec`, but is not initializing the nested `SupplementalGroups`.  And to be honest, I don't know why.

Reply:
No, the returned deployment is the "actual" case. Otherwise, these assertions you added do not test anything, as the object and value here are provided by the tester.
```
	// Check if securityContext was added
	assert.NotNil(t, templateSpec.Spec)
	assert.NotNil(t, templateSpec.Spec.SecurityContext)
	assert.NotNil(t, templateSpec.Spec.SecurityContext.SupplementalGroups)
	assert.Contains(t, templateSpec.Spec.SecurityContext.SupplementalGroups, int64(666))
```
You don't want to remove the SecurityContext lines (starting from line 153) because otherwise the test doesn't pass. But as I said it should really pass without the lines.

Why it doens't pass is actually that the test `pod_test.go` itself has a potential problem and needs to be fixed. Look at the code:
https://github.com/apache/camel-k/blob/main/pkg/trait/pod_test.go#L95-L98
```
	var podSpec v1.PodSpec
	if podSpecTemplate != "" {
		_ = yaml.Unmarshal([]byte(podSpecTemplate), &podSpec)
	}
```
It actually doesn't do any unmarshalling on `podSpec` (check by inserting some println on `podSpec`). It is the root cause of the issue. We need to see why it doesn't as expected and fix it before merging the fix.

Reply:
Removing lines 153-155 and replacing the import on line 25 from `"gopkg.in/yaml.v2"` to `"k8s.io/apimachinery/pkg/util/yaml"` solves the issue.

Reply:
cool. we finally reach the right solution ;-)